lib/libzfs/libzfs_mount.c

```diff
@@ -595,9 +595,13 @@ unmount_one(zfs_handle_t *zhp, const char *mountpoint, int flags)
 		default:
 			libzfs_err = EZFS_UMOUNTFAILED;
 		}
-		return (zfs_error_fmt(zhp->zfs_hdl, libzfs_err,
-		    dgettext(TEXT_DOMAIN, "cannot unmount '%s'"),
-		    mountpoint));
+		if (zhp) {
```

Comment:
If `zhp == NULL` invoking `unmount_one` doesn't make sense at all. Most likely null value is injected from https://github.com/openzfs/zfs/blob/d41856a8996d20c3d1a206a4dfbe10f9373b54f4/lib/libzfs/libzfs_mount.c#L1636 and the if-guard should go there, much like it is done few lines below:
https://github.com/openzfs/zfs/blob/d41856a8996d20c3d1a206a4dfbe10f9373b54f4/lib/libzfs/libzfs_mount.c#L1642
and
https://github.com/openzfs/zfs/blob/d41856a8996d20c3d1a206a4dfbe10f9373b54f4/lib/libzfs/libzfs_mount.c#L1652

Reply:
I mostly put the guard in there because of the comment at https://github.com/openzfs/zfs/blob/d41856a8996d20c3d1a206a4dfbe10f9373b54f4/lib/libzfs/libzfs_mount.c#L1599 suggesting that sometimes calling with a NULL was deliberately permitted - and if you look, do_unmount on Linux doesn't actually use the zhp arg anywhere, so it could have something reasonable happen when called there.

I don't particularly care either way, I just went for the option that changed less behavior.

Reply:
You're right, Rich. Indeed `zhp` is used in neither of OS-specific implementations, what really matters is the `mountpoint`. Quite a surprise to be honest :)